usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_util.c

```diff
@@ -0,0 +1,947 @@
+ * request and therefore possibly pause. If cmd_finish_task() was called on
+ * the interrupt thread a hang condition could occur because IO slots wouldn't
+ * be processed and then freed. So, this portion of the command completion
+ * is run on a taskq.
+ */
+static void
+cmd_finish_task(void *v)
+{
+	pqi_cmd_t	cmd = v;
+	struct scsi_pkt	*pkt;
+
+	pkt = cmd->pc_pkt;
+	if (cmd->pc_poll)
+		sema_v(cmd->pc_poll);
+	if ((pkt->pkt_flags & FLAG_NOINTR) == 0 &&
+	    (pkt->pkt_comp != NULL))
+		(*pkt->pkt_comp)(pkt);
+}
+
+typedef struct qual {
+	int	q_val;
+	char	*q_str;
+} qual_t;
+
+typedef struct code_qual {
+	int	cq_code;
+	qual_t	*cq_list;
+} code_qual_t;
+
+/*
+ * These messages come from pqi2r01 spec section 5.6 table 18.
+ */
+static qual_t pair0[] = { {0, "No error"}, {0, NULL} };
+static qual_t pair1[] = { {0, "Error detected during initialization"},
+	{ 0, NULL } };
+static qual_t pair2[] = { {1, "Invalid PD Function"},
+	{2, "Invalid paramter for PD function"},
+	{0, NULL } };
+static qual_t pair3[] = { {0, "Error creating admin queue pair"},
+	{ 1, "Error delete admin queue pair"},
+	{ 0, NULL} };
+static qual_t pair4[] = { {1, "Invalid IU type in general" },
+	{2, "Invalid IU length in general admin request"},
+	{0, NULL} };
+static qual_t pair5[] = { {1, "Internal error" },
+	{2, "OQ spanning conflict"},
+	{0, NULL} };
+static qual_t pair6[] = { {1, "Error completing PQI soft reset"},
+	{2, "Error completing PQI firmware reset"},
+	{3, "Error completing PQI hardware reset"},
+	{0, NULL} };
+static code_qual_t cq_table[] = {
+	{ 0, pair0 },
+	{ 1, pair1 },
+	{ 2, pair2 },
+	{ 3, pair3 },
+	{ 4, pair4 },
+	{ 5, pair5 },
+	{ 6, pair6 },
+	{ 0, NULL },
+};
+
+static void
+show_error_detail(pqi_state_t s)
+{
+	uint32_t error_reg = G32(s, pqi_registers.device_error);
+	uint8_t		code, qualifier;
+	qual_t		*p;
+	code_qual_t	*cq;
+
+	code = error_reg & 0xff;
+	qualifier = (error_reg >> 8) & 0xff;
+
+	for (cq = cq_table; cq->cq_list != NULL; cq++) {
+		if (cq->cq_code == code) {
+			for (p = cq->cq_list; p->q_str != NULL; p++) {
+				if (p->q_val == qualifier) {
+					cmn_err(CE_NOTE,
+					    "[code=%x,qual=%x]: %s\n",
+					    code, qualifier, p->q_str);
+					return;
+				}
+			}
+		}
+	}
+	cmn_err(CE_NOTE, "Undefined code(%x)/qualifier(%x)\n",
+	    code, qualifier);
+}
+
+/*ARGSUSED*/
+static void
+pqi_catch_release(pqi_io_request_t *io, void *v)
+{
+	/*
+	 * This call can occur if the software times out a command because
+	 * the HBA hasn't responded in the default amount of time, 10 seconds,
+	 * and then the HBA responds. It's occurred a few times during testing
+	 * so catch and ignore.
+	 */
+}
+
+static void
+reinit_io(pqi_io_request_t *io)
+{
+	io->io_cb = pqi_catch_release;
+	io->io_status = 0;
+	io->io_error_info = NULL;
+	io->io_raid_bypass = B_FALSE;
+	io->io_context = NULL;
+	io->io_cmd = NULL;
+}
+
+/* ---- Non-thread safe, for debugging state display code only ---- */
+static char bad_state_buf[64];
```

Comment:
I'm unconvinced doing this unsafely is ok.

Reply:
If only for debug I can live with it.  Maybe this should be behind #ifdef DEBUG tho